File: src/main/java/ru/gdlbo/search/searcher/services/FileService.java
package ru.gdlbo.search.searcher.services;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.jpa.domain.Specification;
import org.springframework.stereotype.Service;
import org.springframework.util.FileCopyUtils;
import ru.gdlbo.search.searcher.config.Config;
import ru.gdlbo.search.searcher.repository.*;

import java.io.File;
import java.io.IOException;
import java.util.List;
import java.util.Optional;

@Service
public class FileService {
    @Autowired
    private FileInfoRepository fileInfoRepository;

    @Autowired
    private FileTempInfoRepository tempInfoRepository;

    @Autowired
    private Config config;
    @Autowired
    private FileTempInfoRepository fileTempInfoRepository;

    public List<FileTempInfo> findAllTemp() {
        return tempInfoRepository.findAll();
    }

    public List<FileTempInfo> findByUser(User user) {
        return tempInfoRepository.findByUser(user);
    }

    public void removeTempFile(Long id) {
        processTempFile(id, true);
    }

    public void approveTempFile(Long id) {
        processTempFile(id, false);
    }

    private void processTempFile(Long id, boolean isRemoval) {
        tempInfoRepository.findById(id).ifPresent(fileInfo -> {
            File tempFile = new File(fileInfo.getLocation());

            if (!tempFile.exists()) {
                throw new IllegalStateException("File " + fileInfo.getLocation() + " does not exist");
            }

            if (!isRemoval) {
                String newPath = config.getPath() + "/" + tempFile.getName();

                try {
                    FileCopyUtils.copy(tempFile, new File(newPath));
                } catch (IOException e) {
                    throw new IllegalStateException("Failed to copy file " + tempFile.getAbsolutePath(), e);
                }

                fileInfo.setLocation(newPath);

                fileInfoRepository.save(new FileInfo(fileInfo));
            }

            if (!tempFile.delete()) {
                throw new IllegalStateException("Failed to delete file " + fileInfo.getLocation());
            }

            tempInfoRepository.delete(fileInfo);
        });
    }

    public List<FileInfo> findFiles(Specification<FileInfo> spec) {
        return fileInfoRepository.findAll(spec);
    }

    public PaginatedResult paginateFileInfos(List<FileInfo> fileInfos, int page) {
        int startIndex = page * 50;
        int endIndex = Math.min(startIndex + 50, fileInfos.size());
        int totalPages = (int) Math.ceil((double) fileInfos.size() / 50);

        List<FileInfo> paginatedFileInfos = fileInfos.subList(startIndex, endIndex);

        return new PaginatedResult(paginatedFileInfos, fileInfos.size() > endIndex, page, totalPages);
    }

    public void saveOrUpdateFile(FileInfo fileInfo) {
        fileInfoRepository.save(fileInfo);
    }

    public void saveTempFile(FileTempInfo fileInfo) {
        tempInfoRepository.save(fileInfo);
    }

    public Optional<FileInfo> getFileById(Long id) {
        return fileInfoRepository.findById(id);
    }

    public Optional<FileTempInfo> getTempFileById(Long id) {
        return fileTempInfoRepository.findById(id);
    }

    public boolean isAnyFilePresent() {
        return fileInfoRepository.count() > 0;
    }

    public boolean doesFileExist(Long id) {
        return fileInfoRepository.existsById(id);
    }

    public boolean existsByDecNumber(String decNumber) {
        return fileInfoRepository.existsByDecNumber(decNumber);
    }
}

File: src/main/java/ru/gdlbo/search/searcher/services/FileService.kt
package ru.gdlbo.search.searcher.services

import org.springframework.beans.factory.annotation.Autowired
import org.springframework.data.jpa.domain.Specification
import org.springframework.stereotype.Service
import org.springframework.util.FileCopyUtils
import ru.gdlbo.search.searcher.config.Config
import ru.gdlbo.search.searcher.repository.PaginatedResult
import ru.gdlbo.search.searcher.repository.files.FileInfo
import ru.gdlbo.search.searcher.repository.files.FileInfoRepository
import ru.gdlbo.search.searcher.repository.files.FileTempInfo
import ru.gdlbo.search.searcher.repository.files.FileTempInfoRepository
import ru.gdlbo.search.searcher.repository.user.User
import java.io.File
import java.util.*
import kotlin.math.ceil
import kotlin.math.min

@Service
class FileService {
    @Autowired
    private val fileInfoRepository: FileInfoRepository? = null

    @Autowired
    private val tempInfoRepository: FileTempInfoRepository? = null

    @Autowired
    private val config: Config? = null

    @Autowired
    private val fileTempInfoRepository: FileTempInfoRepository? = null

    fun findAllTemp(): List<FileTempInfo?> {
        return tempInfoRepository!!.findAll()
    }

    fun findTempByUser(user: User?): List<FileTempInfo?>? {
        return tempInfoRepository!!.findByUser(user)
    }

    fun removeTempFile(id: Long) {
        processTempFile(id, true)
    }

    fun approveTempFile(id: Long) {
        processTempFile(id, false)
    }

    private fun processTempFile(id: Long, isRemoval: Boolean) {
        tempInfoRepository!!.findById(id).ifPresent { fileInfo: FileTempInfo ->
            val tempFile = fileInfo.location?.let { File(it) }
            tempFile?.exists()?.let { check(it) { "File " + fileInfo.location + " does not exist" } }

            if (!isRemoval) {
                val newPath = config?.path + "/" + tempFile?.name

                tempFile?.let { FileCopyUtils.copy(it, File(newPath)) }

                fileInfo.location = newPath

                fileInfoRepository!!.save(FileInfo(fileInfo))
            }

            tempFile?.delete()?.let { check(it) { "Failed to delete file " + fileInfo.location } }
            tempInfoRepository.delete(fileInfo)
        }
    }

    fun findFiles(spec: Specification<FileInfo>): List<FileInfo?> {
        return fileInfoRepository!!.findAll(spec)
    }

    fun paginateFileInfos(fileInfos: List<FileInfo?>, page: Int): PaginatedResult {
        val startIndex = page * 50
        val endIndex = min((startIndex + 50).toDouble(), fileInfos.size.toDouble()).toInt()
        val totalPages = ceil(fileInfos.size.toDouble() / 50).toInt()

        val paginatedFileInfos = fileInfos.subList(startIndex, endIndex).map { it?.toDTO() }

        return PaginatedResult(paginatedFileInfos, fileInfos.size > endIndex, page, totalPages)
    }

    fun saveOrUpdateFile(fileInfo: FileInfo) {
        fileInfoRepository!!.save(fileInfo)
    }

    fun saveTempFile(fileInfo: FileTempInfo) {
        tempInfoRepository!!.save(fileInfo)
    }

    fun getFileById(id: Long): Optional<FileInfo?> {
        return fileInfoRepository!!.findById(id)
    }

    fun getTempFileById(id: Long): Optional<FileTempInfo?> {
        return fileTempInfoRepository!!.findById(id)
    }

    val isAnyFilePresent: Boolean
        get() = fileInfoRepository!!.count() > 0

    fun doesFileExist(id: Long): Boolean {
        return fileInfoRepository!!.existsById(id)
    }

    fun existsByDecNumber(decNumber: String?): Boolean {
        return fileInfoRepository!!.existsByDecNumber(decNumber.toString())
    }
}